torch/csrc/jit/tensorexpr/eval.cpp

```diff
@@ -672,13 +672,75 @@ class SimpleIREvaluatorImpl : public IRVisitor {
     return {};
   }
 
+  void check_bounds_throw(int64_t idx, int64_t bound, const BufPtr& buf) {
+    std::stringstream ss;
+    ss << "Index out of bounds in check_bounds. Index: " << idx
+       << "; bounds: [0, " << bound << ").";
+    throw malformed_input(ss.str(), buf);
+  }
+
+  void check_bounds(const BufPtr& buf, const std::vector<ExprPtr>& indices) {
+    const std::vector<ExprPtr>& dims = buf->dims();
+    if (dims.size() != indices.size()) {
+      // indices are flattened, but not buffer
+      if (indices.size() == 1) {
+        if (dims.size() != buf->strides().size()) {
+          throw malformed_input(
+              "Number of dimensions did not match number of strides", buf);
+        }
+        size_t buf_size = 1;
+        if (dims.size() > 0) {
+          ExprPtr buf_size_expr = immLike(dims[0], 1);
+          ExprPtr negative_one = immLike(dims[0], -1);
+          for (const auto& i : c10::irange(dims.size())) {
+            buf_size_expr = alloc<Add>(
+                buf_size_expr,
+                alloc<Mul>(
+                    alloc<Add>(dims[i], negative_one), buf->strides()[i]));
+          }
+          buf_size_expr->accept(this);
+          buf_size = value().intValue();
+        }
+        indices[0]->accept(this);
+        const auto& index_values = indexVec(value());
+        for (auto& j : index_values) {
+          if (j < 0 || j >= buf_size) {
+            check_bounds_throw(j, buf_size, buf);
+          }
```

Comment:
Hm, shouldn't we be comparing each index vs the sizes of the corresponding dim? I.e. if there is more than one index in index_values, then we shouldn't be comparing it vs flattened buf size. Am I missing something here?

Reply:
This is the case where there the buffer is (n>1)-dimensional, but the index is flattened (i.e. this is the if block for when `dims.size() != indices.size()` and `indices.size() == 1`).

As for why there are multiple indices - I based this off of the implementation of `visit(LoadPtr v)` below, which looks to me like it handles indices as if they can be vectors of indices:

```cpp
    ExprPtr flat_idx =
        flatten_index(v->buf()->dims(), v->indices(), v->buf()->strides());
    flat_idx->accept(this);
    auto index = indexVec(value());
```

So, I'm handling each `index[i]` as if it is a vector of indices. Is this the right way to go about this?

Reply:
I suspect that this vector should always have just one element. Maybe we can assert that (and in the Load too)? But feel free to keep your implementation as-is since indeed that's how Load-visitor is implemented, I don't want us to go into this rabbit hole now :)

Reply:
It looks like sometimes it needs to be handled as a vector (#69961). For example, [Expr.VectorAdd01](https://github.com/pytorch/pytorch/blob/master/test/cpp/tensorexpr/test_expr.cpp#L173-L181).

Reply:
Ah, right, it makes sense when a load/store is vectorized. Thanks for checking this!